### PR TITLE
Runner-State-Loggers refactoring, v20.03

### PR DESCRIPTION
**Pull request by Scitator**
## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] I have read the [Code of Conduct](https://github.com/catalyst-team/catalyst/blob/master/CODE_OF_CONDUCT.md) document.
- [ ] I have read the [Contributing](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md) guide.
- [ ] I have checked the code-style using `make check-codestyle`.
- [ ] I have written tests for all new methods and classes that I created.
- [ ] I have written the docstring in Google format for all the methods and classes that I used.
- [ ] I have checked the docs using `make check-docs`.
- [ ] I have read I need to click 'Login as guest' to see Teamcity build logs.

### CODE REVIEW

File: catalyst/core/callback.py [hunk 1]
```diff
@@ -1,21 +1,27 @@
+from typing import TYPE_CHECKING  # isort:skip
 from enum import IntFlag
-from typing import Callable, List
 
-from catalyst import utils
-from .state import _State
+if TYPE_CHECKING:
```
Comment (reyzharkov) [inline catalyst/core/callback.py:4]: I haven't seen such practice before, but it *really* looks like a hack

File: catalyst/core/callbacks/logging.py [hunk 2]
```diff
@@ -1,19 +1,69 @@
...
 from catalyst import utils
-from catalyst.core import _State, LoggerCallback
+from catalyst.core import _State, Callback, CallbackNode, CallbackOrder
+from catalyst.utils import meters
 from catalyst.utils.tools.tensorboard import SummaryWriter
 from . import formatters
 
 
-class VerboseLogger(LoggerCallback):
+class MetricsManagerCallback(Callback):
```
Comment (reyzharkov) [inline catalyst/core/callbacks/logging.py]: why it is not in metrics.py?

File: catalyst/core/callbacks/metrics.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+class MultiMetricCallback(Callback):
+    """
+    A callback that returns multiple metrics on `state.on_batch_end`
+    """
+    def __init__(
+        self,
+        prefix: str,
+        metric_fn: Callable,
+        list_args: List,
+        input_key: str = "targets",
```
Comment (reyzharkov) [inline catalyst/core/callbacks/metrics.py:121]: It will be great if we can generalise metric/criterion callbacks to have multiple input/output keys

File: catalyst/core/callbacks/optimizer.py [hunk 4]
```diff
@@ -18,12 +14,12 @@ class OptimizerCallback(Callback):
...
-        grad_clip_params: Dict = None,
-        accumulation_steps: int = 1,
-        optimizer_key: str = None,
         loss_key: str = "loss",
+        optimizer_key: str = None,
+        accumulation_steps: int = 1,
+        grad_clip_params: Dict = None,
         decouple_weight_decay: bool = True,
-        save_model_grads: bool = False
+        # save_model_grads: bool = False,
```
Comment (reyzharkov) [inline catalyst/core/callbacks/optimizer.py:25]: todo?

File: catalyst/core/callbacks/timer.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
+from typing import TYPE_CHECKING  # isort:skip
+
+from catalyst.core import Callback, CallbackNode, CallbackOrder
+from catalyst.utils.tools.time_manager import TimeManager
+
+if TYPE_CHECKING:
+    from catalyst.core import _State
+
+
+class TimerCallback(Callback):
```
Comment (reyzharkov) [inline catalyst/core/callbacks/timer.py:5]: This one is really cool! I was waiting for it a long time 💪

File: catalyst/core/callbacks/timer.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    """
+    Logs pipeline execution time
+    """
+    def __init__(self):
+        super().__init__(order=CallbackOrder.Metric + 1, node=CallbackNode.All)
```
Comment (reyzharkov) [inline catalyst/core/callbacks/timer.py:10]: Are you sure about the priority? I believe it should be the very first/last callback at beginning/end as overwise it will miss some over callbacks

File: catalyst/core/callbacks/validation.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
+from typing import TYPE_CHECKING  # isort:skip
+from collections import defaultdict
+
+from catalyst.core import Callback, CallbackNode, CallbackOrder
+
+if TYPE_CHECKING:
+    from catalyst.core import _State
+
+
+class ValidationManagerCallback(Callback):
```
Comment (reyzharkov) [inline catalyst/core/callbacks/validation.py:6]: Why do we need this and why as a separate callback? It will add 2 more lines into yaml configs

Comment (reyzharkov) [reply catalyst/core/callbacks/validation.py:6]: Oh, I see, it's a replacement for `state` callbacks. It makes sense then

Comment (reyzharkov) [reply catalyst/core/callbacks/validation.py:6]: But I hope you will add a docstring

File: catalyst/core/runner.py [hunk 8]
```diff
@@ -235,48 +220,22 @@ def _prepare_for_stage(self, stage: str):
...
-
-        # after callbacks
-        if self.loggers is not None and \
-                (moment == "end" or moment is None):  # for on_exception case
-            for logger in self.loggers.values():
-                getattr(logger, fn_name)(self.state)
-
-        if self.state is not None:
-            getattr(self.state, f"{fn_name}_post")()
+    def _run_event(self, event: str):
```
Comment (reyzharkov) [inline catalyst/core/runner.py:240]: yeah, now it's really simple 😎

File: catalyst/core/runner.py [hunk 9]
```diff
@@ -235,48 +220,22 @@ def _prepare_for_stage(self, stage: str):
...
+        for callback in self.state.callbacks.values():
+            getattr(callback, event)(self.state)
 
     def _batch2device(self, batch: Mapping[str, Any], device: Device):
         output = utils.any2device(batch, device)
         return output
 
     def _run_batch_train_step(self, batch: Mapping[str, Any]):
-        self.state.output = self.forward(batch)
+        self.state.batch_out = self.forward(batch)
```
Comment (reyzharkov) [inline catalyst/core/runner.py:249]: why? (people may be angry as they will have to rewrite *all* of their callbacks)
and I don't think it is more clear

File: catalyst/core/state.py [hunk 10]
```diff
@@ -1,192 +1,133 @@
...
-from catalyst.utils.tools.metric_manager import MetricManager, TimerManager
 from catalyst.utils.tools.typing import (
-    Criterion, Device, Model, Optimizer, Scheduler
+    Criterion, DataLoader, Device, Model, Optimizer, Scheduler
 )
 
+if TYPE_CHECKING:
+    from .callback import Callback  # noqa: F401
+
+STATE_MODEL = Union[Model, Dict[str, Model]]
```
Comment (reyzharkov) [inline catalyst/core/state.py]: why caps and not `StateModel` or something?

File: catalyst/core/callbacks/timer.py (hunk 6 above)
Comment (reyzharkov) [reply catalyst/core/callbacks/timer.py:10]: Another idea - to measure different moments (before & after different callback orders) which will give more inside about which part is the slowest

File: catalyst/core/callbacks/early_stop.py [hunk 11]
```diff
@@ -0,0 +1,59 @@
+from typing import TYPE_CHECKING  # isort:skip
+
+from catalyst.core import Callback, CallbackOrder
+
+if TYPE_CHECKING:
+    from catalyst.core import _State
+
+
+class CheckRunCallback(Callback):
```
Comment (reyzharkov) [inline catalyst/core/callbacks/early_stop.py:9]: Looks like now we will have two add (and remove after) extra lines in config to check if it works.
Maybe you can add this callback as default if `args.check` present on experiment constcurtion phase?

File: catalyst/core/runner.py (hunk 9 above)
Comment (Scitator) [reply catalyst/core/runner.py:249]: deal, what do you think about naming?

File: catalyst/core/callback.py (hunk 1 above)
Comment (Scitator) [reply catalyst/core/callback.py:4]: it prevents cycle imports (for type checking reasons only)
typical practice, btw

File: catalyst/core/state.py (hunk 10 above)
Comment (Scitator) [reply catalyst/core/state.py]: sounds legit

File: catalyst/core/callbacks/early_stop.py (hunk 11 above)
Comment (Scitator) [reply catalyst/core/callbacks/early_stop.py:9]: already

File: catalyst/core/runner.py (hunk 9 above)
Comment (reyzharkov) [reply catalyst/core/runner.py:249]: I think just `input` and `output` are pretty clear and should be OK

File: catalyst/core/callbacks/metrics.py (hunk 3 above)
Comment (Scitator) [reply catalyst/core/callbacks/metrics.py:121]: done

File: catalyst/core/callbacks/metrics.py [hunk 12]
```diff
@@ -50,38 +120,39 @@ def __init__(
...
+class MetricAggregationCallback(Callback):
     """
     A callback to aggregate several metrics in one value.
     """
     def __init__(
         self,
         prefix: str,
         metrics: Union[str, List[str], Dict[str, float]] = None,
-        mode: str = "sum",
+        mode: str = "mean",
```
Comment (reyzharkov) [inline catalyst/core/callbacks/metrics.py:155]: why it has changed?
if we really want to change the default, I think we should raise some alarm warning as it will make the configs from previous versions incompatible

File: catalyst/core/callbacks/validation.py [hunk 13]
```diff
@@ -0,0 +1,44 @@
+from collections import defaultdict
+
+from catalyst.core import _State, Callback, CallbackNode, CallbackOrder
+
+
+class ValidationManagerCallback(Callback):
```
Comment (reyzharkov) [inline catalyst/core/callbacks/validation.py:6]: docstring please

File: catalyst/dl/callbacks/gan.py [hunk 14]
```diff
@@ -53,10 +53,10 @@ def _get(dictionary: dict, keys: Optional[Union[str, List[str]]]) -> Any:
 
     def on_batch_end(self, state: State):
         """On batch end call"""
-        outputs = self._get(state.output, self.output_key)
-        targets = self._get(state.input, self.input_key)
+        outputs = self._get(state.batch_out, self.output_key)
```
Comment (reyzharkov) [inline catalyst/dl/callbacks/gan.py:56]: We can now remove this entire callback (see TODO above - it is now done) and make `WassersteinDistanceCallback` descendant of `MetricCallback`

File: catalyst/core/runner.py [hunk 15]
```diff
@@ -117,6 +103,10 @@ def device(self, value: Device):
                 self._model, "to", device=self._device
             )
 
+    def _init(self):
+        self.experiment: _Experiment = None
+        self.state: _State = None
+
```
Comment (reyzharkov) [inline catalyst/core/runner.py:109]: Maybe we should remove type hints here, because the descendants have exactly the same `_init` except for type hints. Do you think they (type hints) really do matter here?

File: catalyst/core/runner.py [hunk 16]
```diff
@@ -321,88 +274,77 @@ def _run_loader(self, loader: DataLoader):
...
+            if self.state.need_early_stop:
+                self.state.need_early_stop = False
                 break
 
-            self.state.timer.start("_timers/batch_time")
-            self.state.timer.start("_timers/data_time")
-
     def _run_epoch(self, stage: str, epoch: int):
         self._prepare_for_epoch(stage=stage, epoch=epoch)
+        state: _State = self.state
```
Comment (reyzharkov) [inline catalyst/core/runner.py:298]: the same here: remove type hint?

File: catalyst/utils/tools/metric_manager.py [hunk 17]
```diff
@@ -141,4 +112,4 @@ def add_batch_value(
             self._batch_values[metric_name] = metric_value
```
Comment (reyzharkov) [inline catalyst/utils/tools/metric_manager.py:112]: can you add `add_epoch_value` similar to `add_batch_value` for only epoch metrics?

File: catalyst/core/state.py [hunk 18]
```diff
@@ -1,192 +1,143 @@
...
-        # main metrics
-        single_optimizer = isinstance(optimizer, Optimizer)
-        self.lr = None if single_optimizer else defaultdict(lambda: None)
-        self.momentum = None if single_optimizer else defaultdict(lambda: None)
-        self.loss = None
-
-        # data pipeline
-        self.input = None
-        self.output = None
+        # data
```
Comment (reyzharkov) [inline catalyst/core/state.py:168]: it's time to add docstring here

File: catalyst/core/runner.py (hunk 9 above)
Comment (reyzharkov) [reply catalyst/core/runner.py:249]: you can also mark these properties deprecated later

File: catalyst/core/callbacks/metrics.py (hunk 12 above)
Comment (reyzharkov) [reply catalyst/core/callbacks/metrics.py:155]: hmm, as the key also changed (`aggregation_fn` -> `mode`) it's probably OK right now, but still can be misleading

File: catalyst/core/callbacks/early_stop.py [hunk 19]
```diff
@@ -0,0 +1,54 @@
+from catalyst.core import _State, Callback, CallbackOrder
+
+
+class CheckRunCallback(Callback):
+    def __init__(self, num_loader_steps: int = 1, num_epoch_steps: int = 1):
```
Comment (reyzharkov) [inline catalyst/core/callbacks/early_stop.py]: can you return defaults to 3? it was super-secure

File: catalyst/core/callbacks/checkpoint.py [hunk 20]
```diff
@@ -1,61 +1,83 @@
...
+from catalyst.core import _State, Callback, CallbackNode, CallbackOrder
+
+
+def _load_checkpoint(*, filename, state: _State):
+    if os.path.isfile(filename):
+        print(f"=> loading checkpoint {filename}")
+        checkpoint = utils.load_checkpoint(filename)
+
+        if not state.stage_name.startswith("infer"):
+            state.epoch = checkpoint["epoch"] - 1
```
Comment (reyzharkov) [inline catalyst/core/callbacks/checkpoint.py]: it breaks backward compatibility (load older checkpoints), why not start epochs and iterations from 1 instead of 0?

**Comment by reyzharkov** [comment]
.

File: catalyst/core/runner.py (hunk 15 above)
Comment (Scitator) [reply catalyst/core/runner.py:109]: for easier development, it's better to have type hints
just because the have no additional computational overhead during training this trick looks okay :)

File: catalyst/utils/tools/metric_manager.py (hunk 17 above)
Comment (Scitator) [reply catalyst/utils/tools/metric_manager.py:112]: MetricManager is deprecated now... finally
and have no usages in the whole framework

File: catalyst/core/callbacks/checkpoint.py (hunk 20 above)
Comment (Scitator) [reply catalyst/core/callbacks/checkpoint.py]: fixed

File: catalyst/dl/callbacks/gan.py (hunk 14 above)
Comment (Scitator) [reply catalyst/dl/callbacks/gan.py:56]: For now I prefer to leave GAN updated up to you :)
Core refactoring first and all other improvement after that... or it will just never end

File: catalyst/core/runner.py (hunk 9 above)
Comment (Scitator) [reply catalyst/core/runner.py:249]: yup, that why I have introduced additional properties

File: catalyst/core/callbacks/logging.py (hunk 2 above)
Comment (Scitator) [reply catalyst/core/callbacks/logging.py]: fixed

File: catalyst/core/callbacks/optimizer.py (hunk 4 above)
Comment (Scitator) [reply catalyst/core/callbacks/optimizer.py:25]: fixed

File: catalyst/core/callbacks/early_stop.py (hunk 19 above)
Comment (Scitator) [reply catalyst/core/callbacks/early_stop.py]: fixed with env args

File: catalyst/core/callbacks/early_stop.py [hunk 21]
```diff
@@ -0,0 +1,63 @@
+import os
+
+from catalyst.core import _State, Callback, CallbackNode, CallbackOrder
+
+_NUM_LOADER_STEPS = int(os.environ.get("NUM_LOADER_STEPS", 2))
```
Comment (reyzharkov) [inline catalyst/core/callbacks/early_stop.py]: Can you add `CHECK` or `CHECK_RUN` to env variable names, because they may be misleading overwise?

File: catalyst/core/callbacks/metrics.py [hunk 22]
```diff
@@ -165,7 +165,9 @@ def __init__(
                 Must be either ``sum``, ``mean`` or ``weighted_sum``.
             multiplier (float): scale factor for the aggregated metric.
         """
-        super().__init__(CallbackOrder.Metric + 1)
+        super().__init__(
+            order=CallbackOrder.MetricAggregation, node=CallbackNode.All
```
Comment (reyzharkov) [inline catalyst/core/callbacks/metrics.py:169]: why node is not Master?

File: catalyst/core/callbacks/metrics.py [hunk 23]
```diff
@@ -21,7 +21,7 @@ def __init__(
         multiplier: float = 1.0,
         **metrics_kwargs,
     ):
-        super().__init__(CallbackOrder.Metric)
+        super().__init__(order=CallbackOrder.Metric, node=CallbackNode.All)
```
Comment (reyzharkov) [inline catalyst/core/callbacks/metrics.py:24]: master node?

File: catalyst/core/callbacks/metrics.py (hunk 22 above)
Comment (Scitator) [reply catalyst/core/callbacks/metrics.py:169]: because you need to measure metric on all nodes

File: catalyst/core/callbacks/metrics.py (hunk 23 above)
Comment (Scitator) [reply catalyst/core/callbacks/metrics.py:24]: because you need to measure metric on all nodes

File: catalyst/core/callbacks/early_stop.py (hunk 21 above)
Comment (Scitator) [reply catalyst/core/callbacks/early_stop.py]: fixed
